### PR TITLE
GH4391: GHA cross-job download artifact support

### LINKED ISSUE #4391: DownloadArtifacts errors in latest Cake 4.2.0 with 404 error
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched [issues](https://github.com/cake-build/cake/issues) to ensure it has not already been reported

### Cake runner

Cake Frosting

### Cake version

4.2.0

### Operating system

Linux

### Operating system architecture

x64

### CI Server

GitHub Actions

### What are you seeing?

see https://github.com/MonoGame/MonoGame/actions/runs/11728301127/job/32676762557?pr=8561

2024-11-07T19:29:45.1340265Z ========================================
2024-11-07T19:29:45.1341482Z DownloadArtifacts
2024-11-07T19:29:45.1342443Z ========================================
2024-11-07T19:29:45.4873409Z [41m[37;1mAn error occurred when executing task '[0m[41;1m[37;1mDownloadArtifacts[0m[41m[37;1m'.[0m
2024-11-07T19:29:45.4921275Z [41m[37;1mError: [0m[41;1m[37;1mArtifact service call GetSignedArtifactURL failed NotFound (404).[0m
2024-11-07T19:29:45.5987729Z ##[error]Process completed with exit code 255.

### What is expected?

The artifacts should be downloaded.

### Steps to Reproduce

This is the code we use

```
[TaskName("DownloadArtifacts")]
public sealed class DownloadArtifactsTask : AsyncFrostingTask<BuildContext>
{
    public override bool ShouldRun(BuildContext context) => context.BuildSystem().IsRunningOnGitHubActions;

    public override async Task RunAsync(BuildContext context)
    {
        context.CreateDirectory("nugets");
        foreach (var os in new[] { "windows", "macos", "linux" })
            await context.GitHubActions().Commands.DownloadArtifact($"nuget-{os}", "nugets");

        context.CreateDirectory("vsix");
        await context.GitHubActions().Commands.DownloadArtifact("MonoGame.Templates.VSExtension.vsix", "vsix");
    }
}
```

### Output log

2024-11-07T19:29:23.8793616Z Current runner version: '2.320.0'
2024-11-07T19:29:23.8817759Z ##[group]Operating System
2024-11-07T19:29:23.8818362Z Ubuntu
2024-11-07T19:29:23.8818835Z 22.04.5
2024-11-07T19:29:23.8819195Z LTS
2024-11-07T19:29:23.8819542Z ##[endgroup]
2024-11-07T19:29:23.8820000Z ##[group]Runner Image
2024-11-07T19:29:23.8820423Z Image: ubuntu-22.04
2024-11-07T19:29:23.8820823Z Version: 20241103.1.0
2024-11-07T19:29:23.8821900Z Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20241103.1/images/ubuntu/Ubuntu2204-Readme.md
2024-11-07T19:29:23.8823366Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu22%2F20241103.1
2024-11-07T19:29:23.8824237Z ##[endgroup]
2024-11-07T19:29:23.8824688Z ##[group]Runner Image Provisioner
2024-11-07T19:29:23.8825174Z 2.0.384.1
2024-11-07T19:29:23.8825521Z ##[endgroup]
2024-11-07T19:29:23.8840065Z ##[group]GITHUB_TOKEN Permissions
2024-11-07T19:29:23.8841818Z Contents: write
2024-11-07T19:29:23.8842258Z Metadata: read
2024-11-07T19:29:23.8842970Z Packages: write
2024-11-07T19:29:23.8843466Z ##[endgroup]
2024-11-07T19:29:23.8846851Z Secret source: Actions
2024-11-07T19:29:23.8847464Z Prepare workflow directory
2024-11-07T19:29:23.9460552Z Prepare all required actions
2024-11-07T19:29:23.9616597Z Getting action download info
2024-11-07T19:29:24.1488138Z Download action repository 'actions/checkout@v4' (SHA:11bd71901bbe5b1630ceea73d27597364c9af683)
2024-11-07T19:29:24.2556203Z Download action repository 'actions/setup-dotnet@v4' (SHA:3e891b0cb619bf60e2c25674b222b8940e2c1c25)
2024-11-07T19:29:24.5820686Z Download action repository 'crazy-max/ghaction-github-runtime@v3' (SHA:b3a9207c0e1ef41f4cf215303c976869d0c2c1c4)
2024-11-07T19:29:24.8544162Z Download action repository 'ncipollo/release-action@v1' (SHA:2c591bcc8ecdcd2db72b97d6147f871fcd833ba5)
2024-11-07T19:29:25.1540575Z Complete job name: Deploy
2024-11-07T19:29:25.2348187Z ##[group]Run actions/checkout@v4
2024-11-07T19:29:25.2348659Z with:
2024-11-07T19:29:25.2348958Z   repository: MonoGame/MonoGame
2024-11-07T19:29:25.2349580Z   token: ***
2024-11-07T19:29:25.2349879Z   ssh-strict: true
2024-11-07T19:29:25.2350308Z   ssh-user: git
2024-11-07T19:29:25.2350630Z   persist-credentials: true
2024-11-07T19:29:25.2350986Z   clean: true
2024-11-07T19:29:25.2351292Z   sparse-checkout-cone-mode: true
2024-11-07T19:29:25.2351696Z   fetch-depth: 1
2024-11-07T19:29:25.2351980Z   fetch-tags: false
2024-11-07T19:29:25.2352283Z   show-progress: true
2024-11-07T19:29:25.2352602Z   lfs: false
2024-11-07T19:29:25.2352864Z   submodules: false
2024-11-07T19:29:25.2353170Z   set-safe-directory: true
2024-11-07T19:29:25.2353512Z ##[endgroup]
2024-11-07T19:29:25.4649180Z Syncing repository: MonoGame/MonoGame
2024-11-07T19:29:25.4650930Z ##[group]Getting Git version info
2024-11-07T19:29:25.4651663Z Working directory is '/home/runner/work/MonoGame/MonoGame'
2024-11-07T19:29:25.4652635Z [command]/usr/bin/git version
2024-11-07T19:29:25.4687861Z git version 2.47.0
2024-11-07T19:29:25.4712435Z ##[endgroup]
2024-11-07T19:29:25.4736703Z Temporarily overriding HOME='/home/runner/work/_temp/9072f744-05f7-4b5a-b2b5-52a229483980' before making global git config changes
2024-11-07T19:29:25.4739041Z Adding repository directory to the temporary git global config as a safe directory
2024-11-07T19:29:25.4742992Z [command]/usr/bin/git config --global --add safe.directory /home/runner/work/MonoGame/MonoGame
2024-11-07T19:29:25.4778363Z Deleting the contents of '/home/runner/work/MonoGame/MonoGame'
2024-11-07T19:29:25.4782078Z ##[group]Initializing the repository
2024-11-07T19:29:25.4786827Z [command]/usr/bin/git init /home/runner/work/MonoGame/MonoGame
2024-11-07T19:29:25.4884385Z hint: Using 'master' as the name for the initial branch. This default branch name
2024-11-07T19:29:25.4885496Z hint: is subject to change. To configure the initial branch name to use in all
2024-11-07T19:29:25.4886704Z hint: of your new repositories, which will suppress this warning, call:
2024-11-07T19:29:25.4887343Z hint:
2024-11-07T19:29:25.4887780Z hint: 	git config --global init.defaultBranch <name>
2024-11-07T19:29:25.4888635Z hint:
2024-11-07T19:29:25.4889156Z hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
2024-11-07T19:29:25.4890092Z hint: 'development'. The just-created branch can be renamed via this command:
2024-11-07T19:29:25.4890739Z hint:
2024-11-07T19:29:25.4891039Z hint: 	git branch -m <name>
2024-11-07T19:29:25.4912842Z Initialized empty Git repository in /home/runner/work/MonoGame/MonoGame/.git/
2024-11-07T19:29:25.4924196Z [command]/usr/bin/git remote add origin https://github.com/MonoGame/MonoGame
2024-11-07T19:29:25.4958518Z ##[endgroup]
2024-11-07T19:29:25.4959505Z ##[group]Disabling automatic garbage collection
2024-11-07T19:29:25.4963688Z [command]/usr/bin/git config --local gc.auto 0
2024-11-07T19:29:25.4993695Z ##[endgroup]
2024-11-07T19:29:25.4994701Z ##[group]Setting up auth
2024-11-07T19:29:25.5001515Z [command]/usr/bin/git config --local --name-only --get-regexp core\.sshCommand
2024-11-07T19:29:25.5033255Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :"
2024-11-07T19:29:25.5372023Z [command]/usr/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2024-11-07T19:29:25.5403282Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :"
2024-11-07T19:29:25.5632521Z [command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
2024-11-07T19:29:25.5667754Z ##[endgroup]
2024-11-07T19:29:25.5668955Z ##[group]Fetching the repository
2024-11-07T19:29:25.5678180Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +1b1c785ef9a9c3baf1388b0eb416fca80208df87:refs/remotes/origin/dev/dellis1972/cake-frosting
2024-11-07T19:29:26.8950470Z From https://github.com/MonoGame/MonoGame
2024-11-07T19:29:26.8951874Z  * [new ref]         1b1c785ef9a9c3baf1388b0eb416fca80208df87 -> origin/dev/dellis1972/cake-frosting
2024-11-07T19:29:26.8976703Z ##[endgroup]
2024-11-07T19:29:26.8977460Z ##[group]Determining the checkout info
2024-11-07T19:29:26.8979249Z ##[endgroup]
2024-11-07T19:29:26.8984317Z [command]/usr/bin/git sparse-checkout disable
2024-11-07T19:29:26.9026705Z [command]/usr/bin/git config --local --unset-all extensions.worktreeConfig
2024-11-07T19:29:26.9055499Z ##[group]Checking out the ref
2024-11-07T19:29:26.9060379Z [command]/usr/bin/git checkout --progress --force -B dev/dellis1972/cake-frosting refs/remotes/origin/dev/dellis1972/cake-frosting
2024-11-07T19:29:27.3280740Z Switched to a new branch 'dev/dellis1972/cake-frosting'
2024-11-07T19:29:27.3283444Z branch 'dev/dellis1972/cake-frosting' set up to track 'origin/dev/dellis1972/cake-frosting'.
2024-11-07T19:29:27.3307017Z ##[endgroup]
2024-11-07T19:29:27.3346182Z [command]/usr/bin/git log -1 --format=%H
2024-11-07T19:29:27.3369711Z 1b1c785ef9a9c3baf1388b0eb416fca80208df87
2024-11-07T19:29:27.3673501Z ##[group]Run actions/setup-dotnet@v4
2024-11-07T19:29:27.3673868Z with:
2024-11-07T19:29:27.3674104Z   dotnet-version: 8.0.x
2024-11-07T19:29:27.3674377Z   cache: false
2024-11-07T19:29:27.3674619Z ##[endgroup]
2024-11-07T19:29:27.5300655Z [command]/home/runner/work/_actions/actions/setup-dotnet/v4/externals/install-dotnet.sh --skip-non-versioned-files --runtime dotnet --channel LTS
2024-11-07T19:29:27.7777575Z dotnet-install: .NET Core Runtime with version '8.0.10' is already installed.
2024-11-07T19:29:27.7805965Z [command]/home/runner/work/_actions/actions/setup-dotnet/v4/externals/install-dotnet.sh --skip-non-versioned-files --channel 8.0
2024-11-07T19:29:28.0756346Z dotnet-install: .NET Core SDK with version '8.0.403' is already installed.
2024-11-07T19:29:28.0896354Z ##[group]Run crazy-max/ghaction-github-runtime@v3
2024-11-07T19:29:28.0896988Z with:
2024-11-07T19:29:28.0897691Z   github-token: ***
2024-11-07T19:29:28.0898074Z env:
2024-11-07T19:29:28.0898473Z   DOTNET_ROOT: /usr/share/dotnet
2024-11-07T19:29:28.0898834Z ##[endgroup]
2024-11-07T19:29:28.1266007Z ACTIONS_RUNNER_ACTION_ARCHIVE_CACHE=/opt/actionarchivecache
2024-11-07T19:29:28.1275241Z ACTIONS_RUNTIME_URL=https://pipelinesghubeus4.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1327770Z ACTIONS_RUNTIME_TOKEN=***
2024-11-07T19:29:28.1328719Z ACTIONS_CACHE_URL=https://acghubeus1.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1329917Z ACTIONS_RESULTS_URL=https://results-receiver.actions.githubusercontent.com/
2024-11-07T19:29:28.1401020Z ##[group]Run dotnet run --project build/Build.csproj -- --target=Deploy
2024-11-07T19:29:28.1401812Z [36;1mdotnet run --project build/Build.csproj -- --target=Deploy[0m
2024-11-07T19:29:28.1489312Z shell: /usr/bin/bash -e {0}
2024-11-07T19:29:28.1489971Z env:
2024-11-07T19:29:28.1490353Z   DOTNET_ROOT: /usr/share/dotnet
2024-11-07T19:29:28.1490873Z   ACTIONS_RUNNER_ACTION_ARCHIVE_CACHE: /opt/actionarchivecache
2024-11-07T19:29:28.1491932Z   ACTIONS_RUNTIME_URL: https://pipelinesghubeus4.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1532752Z   ACTIONS_RUNTIME_TOKEN: ***
2024-11-07T19:29:28.1533711Z   ACTIONS_CACHE_URL: https://acghubeus1.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1534762Z   ACTIONS_RESULTS_URL: https://results-receiver.actions.githubusercontent.com/
2024-11-07T19:29:28.1535498Z   GITHUB_TOKEN: ***
2024-11-07T19:29:28.1536052Z   NUGET_API_KEY: ***
2024-11-07T19:29:28.1536556Z   MARKETPLACE_PAT: ***
2024-11-07T19:29:28.1536979Z ##[endgroup]
2024-11-07T19:29:38.0199827Z /usr/share/dotnet/sdk/8.0.403/Microsoft.Common.CurrentVersion.targets(2412,5): warning MSB3246: Resolved file has a bad image, no metadata, or is otherwise inaccessible. PE image does not have metadata. PE image does not have metadata. [/home/runner/work/MonoGame/MonoGame/build/Build.csproj]
2024-11-07T19:29:45.0515620Z Version: 3.8.2.1570-develop
2024-11-07T19:29:45.0516505Z RepositoryUrl: https://github.com/MonoGame/MonoGame
2024-11-07T19:29:45.0517337Z BuildConfiguration: Release
2024-11-07T19:29:45.1339392Z 
2024-11-07T19:29:45.1340265Z ========================================
2024-11-07T19:29:45.1341482Z DownloadArtifacts
2024-11-07T19:29:45.1342443Z ========================================
2024-11-07T19:29:45.4873409Z [41m[37;1mAn error occurred when executing task '[0m[41;1m[37;1mDownloadArtifacts[0m[41m[37;1m'.[0m
2024-11-07T19:29:45.4921275Z [41m[37;1mError: [0m[41;1m[37;1mArtifact service call GetSignedArtifactURL failed NotFound (404).[0m
2024-11-07T19:29:45.5987729Z ##[error]Process completed with exit code 255.
2024-11-07T19:29:45.6110234Z Post job cleanup.
2024-11-07T19:29:45.7043285Z [command]/usr/bin/git version
2024-11-07T19:29:45.7081017Z git version 2.47.0
2024-11-07T19:29:45.7122530Z Temporarily overriding HOME='/home/runner/work/_temp/2675194b-04af-46b8-b5cf-45c3bc204a88' before making global git config changes
2024-11-07T19:29:45.7123826Z Adding repository directory to the temporary git global config as a safe directory
2024-11-07T19:29:45.7135748Z [command]/usr/bin/git config --global --add safe.directory /home/runner/work/MonoGame/MonoGame
2024-11-07T19:29:45.7169759Z [command]/usr/bin/git config --local --name-only --get-regexp core\.sshCommand
2024-11-07T19:29:45.7200768Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :"
2024-11-07T19:29:45.7487886Z [command]/usr/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2024-11-07T19:29:45.7511842Z http.https://github.com/.extraheader
2024-11-07T19:29:45.7523687Z [command]/usr/bin/git config --local --unset-all http.https://github.com/.extraheader
2024-11-07T19:29:45.7556284Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :"
2024-11-07T19:29:45.8021865Z Cleaning up orphan processes
2024-11-07T19:29:45.8422109Z Terminate orphan process: pid (1929) (dotnet)

### PR DESCRIPTION
**Pull request by devlead**
Enables support for cross-job download of artifacts within the current run, by fetching job run id for specified artifact.

* fixes #4391

### CODE REVIEW

**Comment by dellis1972** [comment]
Looks ok to me, but then again I'm not familiar with the codebase 🤣

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -34,7 +45,7 @@ jobs:
 
       - name: Run Cake script
         id: build-cake
-        uses: cake-build/cake-action@v1
+        uses: cake-build/cake-action@master
```
Comment (gep13) [inline .github/workflows/build.yml:48]: Is there a reason that we can't pin to the version/tag of the action?

**Review by gep13 — APPROVED** [review]
LGTM!

File: .github/workflows/build.yml (hunk 1 above)
Comment (dellis1972) [reply .github/workflows/build.yml:48]: That will be the next question :D which release will this get into?

Comment (devlead) [reply .github/workflows/build.yml:48]: @dellis1972  This PR will be released as part of 5.0 release, planned shortly (as son as we get the cycles to do so) after .NET 9 release tomorrow.

You could already now try the pre-release via our Azure Artifacts feed.
https://dev.azure.com/cake-build/Cake/_artifacts/feed/cake/NuGet/Cake.Frosting/overview/5.0.0-alpha0030
```xml
  <PackageReference Include="Cake.Frosting" Version="5.0.0-alpha0030" />
```

if you want to test in your PR now, you can add the public source as
```xml
<add key="cake" value="https://pkgs.dev.azure.com/cake-build/Cake/_packaging/cake/nuget/v3/index.json" />
```

**Pull request merged**